library/alloc/src/slice.rs

```diff
@@ -900,23 +903,22 @@ where
             // If `is_less` panics at any point during the process, `hole` will get dropped and
             // fill the hole in `v` with `tmp`, thus ensuring that `v` still holds every object it
             // initially held exactly once.
-            let mut hole = InsertionHole { src: &mut *tmp, dest: &mut v[1] };
-            ptr::copy_nonoverlapping(&v[1], &mut v[0], 1);
-
-            for i in 2..v.len() {
-                if !is_less(&v[i], &*tmp) {
-                    break;
-                }
-                ptr::copy_nonoverlapping(&v[i], &mut v[i - 1], 1);
-                hole.dest = &mut v[i];
+            let mut hole = InsertionHole { src: &*tmp, dest: v.get_unchecked_mut(end - 1) };
+            ptr::copy_nonoverlapping(hole.dest, v.get_unchecked_mut(end), 1);
+
+            let mut i = end - 1;
```

Comment:
`NonZeroUsize` + `while let` would express the semantics better.

Reply:
I don't understand.  `i` can be `0`

Reply:
If it's zero, the loop will not be entered as usual (it would be `None`).

Reply:
I see...  I think stylistically it is very clear as written, the NonZeroUsize adds additional cognitive overhead?

I'm not clear on the performance implications and also it would require a nightly feature:
![image](https://user-images.githubusercontent.com/6633152/140380961-1e45de02-d1a4-4af3-a23d-2c20792cd604.png)

Reply:
You are right about the nightliness of addition, I forgot about that.
Checked addition is slower and I wouldn't advise that here, but I thought saturating addition isn't necessarily, by the way.

Yeah, I see how some programmers will find `NonZeroUsize` odd as it stands out. On the other hand, some will like the correctness by construction aspect and informativeness.

Reply:
Just use any nightly features if you want. The Standard Library is always built by the latest nightly compiler.

Reply:
Thanks.  I believe it is best to leave the `while i > 0` loop as is.  `NonZeroUsize` would require more code and wouldn't necessarily be clearer.